[ASCII-2580] Implement function to safely fetch or create artifacts concurrently

misteriaud opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?
This PR improves the mechanism for creating and fetching authentication artifacts used for inter-process communication between Agents. It introduces a more robust and fault-tolerant approach that eliminates race conditions, maintains process independence, and ensures reliable artifact retrieval.

### Motivation
The existing authentication mechanism relies on a strict startup order where the coreAgent generates artifacts before subAgents start. Initially, this worked well because subAgents acted only as clients. However, with subAgents now also acting as servers, this assumption has become problematic.

In reality, agent processes do not follow a guaranteed startup sequence, leading to race conditions when subAgents attempt to load authentication artifacts before they are created by the coreAgent. This results in failures, inconsistent behaviors, and unnecessary complexity.

Several constraints make this problem particularly challenging:
- **Configurable artifact paths**: Authentication artifact locations are configurable, preventing pre-generation at installation.
- **Independent startup model**: The system was designed so that Agents can start independently without strict dependencies. In host-based deployments, the coreAgent is historically the last to start, making it even more critical for subAgents to handle authentication artifact retrieval reliably.
- **Robustness and fault tolerance**: Simple retries or waiting for the coreAgent would introduce tighter coupling between processes, reducing flexibility and making the system more fragile.

Previous workarounds, such as lazy-loading or background fetching, have led to unreliable behavior and inconsistent patterns. A more structured approach is necessary to ensure subAgents can reliably obtain authentication artifacts without unnecessary dependencies.

### Implementation Description

1. **Optimistic Fetching:**
   - Attempts to retrieve the artifact from disk.
   - If valid, it is returned immediately.

2. **Lock-Based Creation:**
   - Uses a temporary file to prevent partial artifacts if creation fails.
   - If the artifact is missing, a file-based write lock is acquired.
   - Processes can independently verify if the artifact is valid, avoiding read locks.

3. **Retry-Based Fetching:**
   - If another process holds the lock, retries until the artifact is available.
   - Ensures reliable retrieval without race conditions.

### Describe how you validated your changes

I added unit tests covering the core mechanism. e2e tests involving CoreAgent and TraceAgent ensure real-world reliability, as they are currently the only processes authorized to create `auth_token`.

### Possible Drawbacks / Trade-offs

This mechanism relies on OS-level file locking. In worst-case scenarios, Agent initialization may be delayed if another process locks the file and takes time to release it.

### Additional Notes

#### Kubernetes Compatibility
This mechanism works well in Kubernetes, where Agents run in separate containers but share a volume for authentication artifacts:

- **coreAgent** has exclusive write permissions.
- **SubAgents** must wait for coreAgent to generate the artifact, relying on the retry-based fetching loop.
- Maintains startup flexibility without missing authentication data.

ogaca-dd commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,191 @@
+	return res, err
+}
+
+func createTmpFile(location string) (*os.File, error) {
```
Did you consider using https://pkg.go.dev/os#CreateTemp

misteriaud replied:
Good idea ! Change done in cb940c6d3fc17d84b15b0129c6ba242de20e09ae.

pgimalac commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,169 @@
+	fileLock := flock.New(location + ".lock")
+	defer func() {
+		log.Debugf("releasing lock for file %v", location)
+		fileLock.Close()
+		if err := os.Remove(fileLock.Path()); err != nil && os.IsNotExist(err) {
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏Is there a race here ? You delete the lock file while some other process could be holding it

misteriaud replied:
Indeed ! I removed the `os.Remove` call to avoid this. But I wonder if keeping `.lock` files could cause any problems ?

pgimalac replied:
As long as you don't keep it locked it should be ok I guess

pgimalac commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,169 @@
+		}
+	}()
+
+	// trying to lock artifact file
+	err = backoff.Retry(func() error {
+		log.Tracef("trying to lock file %v", fileLock.Path())
+		locked, err := fileLock.TryLock()
+
+		if err != nil {
+			return backoff.Permanent(fmt.Errorf("unable to TryLock: %v", err.Error()))
+		}
+
+		if !locked {
+			return fmt.Errorf("unable to acquire lock %v", location)
+		}
+		return nil
+	}, factory.BackoffTimer())
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏Why not just use [TryLockContext](https://pkg.go.dev/github.com/gofrs/flock#Flock.TryLockContext) ?

GustavoCaso replied:
That way we can get rid of the `backoff` dependency, we could still configure the `retryDelay ` per factory

pgimalac replied:
To be fair backoff is already used by every agent binary, but yeah it's just simpler

GustavoCaso replied:
note: remove dependency from the filesystem package 😉

misteriaud replied:
Good idea ! I'll do that

pgimalac commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,169 @@
+// if it is initially unsuccessful.
+//
+// This function is thread-safe and non-blocking.
+func FetchOrCreateArtifact[T any](factory ArtifactBuilder[T]) (T, error) {
```
<!-- Warning represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 warning:** ‏I think there is a big race here, ie:
- process 1 fails to fetch
- process 2 fails to fetch
- process 1 locks, create the artifact, unlocks
- process 2 locks, create the artifact (which will override the previous one because we use `os.Rename`), unlocks

Both processes end up with different artifacts...

I think you would have to lock when fetching initially to fix it ?

pgimalac replied:
Nevermind I somehow missed that you re-try fetching before writing ! 🤦

GustavoCaso commented on pkg/api/security/security.go:
```diff
@@ -196,37 +229,12 @@ func getClusterAgentAuthToken(config configModel.Reader, tokenCreationAllowed bo
-	if err != nil {
-		return "", fmt.Errorf("empty cluster_agent.auth_token and cannot read %s: %s", tokenAbsPath, err)
+	factory := &dcaAuthTokenFactory{config}
+	log.Debugf("Empty cluster_agent.auth_token, loading from %s", factory.Location())
```
Should we still print this line? It seems that we print all the time no matter if we fetch or create or just fetch. 

Should we print it only if is truly empty after we fetch it?

misteriaud replied:
I think it's important to keep the same logs as printed before. 

> Should we print it only if is truly empty after we fetch it?
The clusterAgent first try to get the authToken value from the configuration, so this log is explicitly to state that it haven't been loaded from it. It's kind of a first fetch.

GustavoCaso commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,169 @@
+	return res, err
+}
+
+func generateTmpArtifact[T any](factory ArtifactBuilder[T]) (T, string, error) {
```
nitpick: I would move this method after `FetchOrCreateArtifact`. That way the order of the methods are public and then private 😄 I forces the reader to read the file from top to bottom

GustavoCaso commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,169 @@
+
+	// trying to lock artifact file
+	err = backoff.Retry(func() error {
+		log.Tracef("trying to lock file %v", fileLock.Path())
```
Is the use of `Tracef` intentional?

GustavoCaso commented:
Could I suggest that rather than using  a third party library for the locking mechanism instead we use OS native calls? 
Linux has the `flock` and Windows has `LockFileEx`. We can even have a look at how Go does internally 
https://cs.opensource.google/go/go/+/refs/tags/go1.23.5:src/cmd/go/internal/lockedfile/internal/filelock/filelock.go

misteriaud commented on pkg/api/security/cert/cert_getter.go:
```diff
@@ -37,92 +37,60 @@ func GetCertFilepath(config configModel.Reader) string {
-// Requires that the config has been set up before calling
-func FetchAgentIPCCert(config configModel.Reader) ([]byte, []byte, error) {
-	return fetchAgentIPCCert(config, false)
+type certifaceFactory struct {
```
Indeed ! It wasn't on purpose 😅

misteriaud committed:
go mod tidy the repo

misteriaud committed:
implement function to safely load or generate artifacts in parallel

misteriaud committed:
update auth_token and IPC cert code to use the new concurrent write function

misteriaud committed:
update SetAuthToken and CreateOrSetAuthToken from pkg/api/util

misteriaud committed:
update all related unit tests

misteriaud committed:
update minor things

ogaca-dd commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,158 @@
+
+	if err != nil || !locked {
+		if err == nil {
+			err = fmt.Errorf("unknown error")
```
Are `err ==nil` and `locked != true` when the  timeout exceed? If yes the error message can reflect that.

pgimalac replied:
Yeah here either you know the error, or you had a timeout, so this could be clearer

misteriaud replied:
When the context timeout `err != nil`. It is tested in `TestContextCancellation` test, the error contains `context deadline exceeded`.

pgimalac replied:
Really ? How can you have `err == nil` and `locked == false` then ?

misteriaud replied:
I don't know specific case where it could happen, but I think it's important to not imply that `err == nil` means `locked == true` IMO.

misteriaud replied:
> Yeah I agree, it's just surprising for the API of the lib 😄
Same !

ogaca-dd commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,158 @@
+	// Happy path is to be able to generate and store artifact
+	// We prefer generating a temporary artifact and moving it to its final location
+	// to avoid having a half written artifact in case of a failure
+	createdArtifact, tmpLocation, err := generateTmpArtifact(location, factory)
```
Is there a particular reason to run this code after trying to get the lock?
It will make the cleanup of `tmpLocation` easier as you don't have to handle it for each `return`.

misteriaud replied:
Good point I switched it the other way around. It makes the code cleaner and it's not a big trade-off considering that this function will be called only once at Agent start.

ogaca-dd commented on pkg/util/filesystem/concurrent_write_test.go:
```diff
@@ -0,0 +1,164 @@
+
+func TestFetchArtifact(t *testing.T) {
+	t.Parallel()
+	location, mockFactory := newMockArtiFactory(t)
```
Is it expected to use the same value for `location` accros tests  and use `t.Parallel`? You might have race condition.

misteriaud replied:
`newMockArtiFactory(t)` return a different location every time, so every instance is individual.

ogaca-dd commented on pkg/util/filesystem/concurrent_write_test.go:
```diff
@@ -0,0 +1,164 @@
+	location, mockFactory := newMockArtiFactory(t)
+
+	// Ensure the artifact file does not exist
+	os.Remove(location)
```
Because of `t.Parallel()`, you may have another test to create this file.

misteriaud replied:
Same as previous comment, since `newMockArtiFactory(t)` returns a unique location, their is no chance that tests are overlapping.

pgimalac commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,169 @@
+
+// FetchArtifact attempts to load an artifact using the provided factory.
+// If the artifact does not exist, it return an error.
+func FetchArtifact[T any](factory ArtifactBuilder[T]) (T, error) {
```
You can still pass the same object, I just meant that the function can accept a smaller interface (which makes it easier to use)

pgimalac commented on pkg/api/security/security.go:
```diff
@@ -123,50 +143,35 @@ func GetAuthTokenFilepath(config configModel.Reader) string {
-		if e != nil {
-			return "", fmt.Errorf("can't create agent authentication token value: %s", e)
-		}
+type dcaAuthTokenFactory struct {
+}
```
Is this type different from `authtokenFactory` ?

misteriaud replied:
It was different the interface had `Location()` method. Know it can be merge, so I did it. Good idea !

pgimalac commented on pkg/api/util/util.go:
```diff
@@ -16,13 +17,16 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
+const createAndSetAuthTokenTimeout = 10 * time.Second
```
Could you make that configurable ?

misteriaud replied:
Yes totally, I dit it in a5329476283608da402c04ca36e6961eff4c188c.

pgimalac commented on pkg/api/util/util.go:
```diff
@@ -104,14 +108,41 @@ func CreateAndSetAuthToken(config model.Reader) error {
+	errs := make(chan error, 2)
+	var ipccert, ipckey []byte
+
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		var err error
+		token, err = pkgtoken.FetchOrCreateAuthToken(ctx, config)
+		if err != nil {
+			err = fmt.Errorf("error while creating or fetching auth token: %v", err.Error())
+		}
+		errs <- err
+	}()
+
+	go func() {
+		defer wg.Done()
+		var err error
+		ipccert, ipckey, err = cert.FetchOrCreateIPCCert(ctx, config)
+		if err != nil {
+			err = fmt.Errorf("error while creating or fetching IPC cert: %v", err.Error())
+		}
+		errs <- err
+	}()
```
I think not but I'd rather ask, would it be an issue if one process created the auth token and another one the ipc cert ?

misteriaud replied:
there's no problem with 2 different processes each creating an artifact. We just need that they use the same one at the end of the process.

pgimalac commented on pkg/api/util/util.go:
```diff
@@ -139,7 +170,7 @@ func CreateAndSetAuthToken(config model.Reader) error {
 func IsInitialized() bool {
 	tokenLock.RLock()
-	defer tokenLock.Unlock()
+	defer tokenLock.RUnlock()
```
Huum does that fix need to be backported ?
I'm not sure what could be the impact of calling `Unlock` when read-locked 😕

misteriaud replied:
This function wasn't used before this PR, so no need to backport it.
The race detector was making my test fail because of it. Maybe it can lead to deadlock, I don't know...

pgimalac commented on pkg/api/util/util.go:
```diff
@@ -186,8 +217,11 @@ func InitDCAAuthToken(config model.Reader) error {
 		return nil
 	}
 
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
You should probably use the constant here ? (and if you do make it configurable then use the config)

pgimalac commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,158 @@
+// FetchOrCreateArtifact attempts to load an artifact using the provided factory.
+// If the artifact does not exist, it generates a new one, stores it, and returns it.
+//
+// The function first tries to load the artifact using the factory's location.
```
Might need to update this sentence

pgimalac commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,158 @@
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// ArtifactBuilder is a generic interface for building, serializing, and deserializing artifacts.
+// The type parameter T represents the in-memory type of the artifact.
+type ArtifactBuilder[T any] interface {
+	Generate() (T, error)
+	Serialize(T) ([]byte, error)
+	Deserialize([]byte) (T, error)
+}
```
I feel like `Generate` will always just be used to get a value to pass to `Serialize`, if so is there any reason to split the methods here ? It could just be eg. `GenerateSerialized`

misteriaud replied:
Good idea, done in 95ecd938d28c8fe2cc974d36b969b6de4ac8988e.

pgimalac commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,158 @@
+		if err == nil {
+			err = fmt.Errorf("unknown error")
+		}
+		if !locked {
```
I think here you'll always be unlocked (in case of error the lock should stay open)

pgimalac commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,158 @@
+	}()
+
+	// trying to lock artifact file
+	locked, err := fileLock.TryLockContext(ctx, 100*time.Millisecond)
```
I believe it will try locking every 100ms until the context is done, it's not a timeout before stopping

misteriaud replied:
Good point, I didn't think a lot about what duration could be appropriate. I switched to 500ms.

pgimalac commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,158 @@
+		return zero, fmt.Errorf("unable to generate temporary artifact: %v", err.Error())
+	}
+
+	fileLock := flock.New(location + ".lock")
+	defer func() {
+		log.Debugf("releasing lock for file %v", location)
+		err := fileLock.Unlock()
```
Are you sure it's safe to call `Unlock` without holding the lock ? If not you should move that after locking

misteriaud replied:
Good question ! It does nothing if the file is not lock: 

```golang
	// If we aren't locked or if the lockfile instance is nil
	// just return a nil error because we are unlocked.
	if (!f.l && !f.r) || f.fh == nil {
		return nil
	}
```

[link to the source code of flock](https://github.com/gofrs/flock/blob/v0.12.1/flock_unix.go#L98-L102)

pgimalac replied:
Could you mention that explicitly in a comment ?

pgimalac commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,158 @@
+		return zero, "", fmt.Errorf("unable to serialize new artifact: %v", err)
+	}
+
+	perms, err := NewPermission()
+	if err != nil {
+		return zero, "", fmt.Errorf("unable to prepare artifact: %v", err)
+	}
+
+	tmpFile, err := os.CreateTemp(filepath.Dir(location), "tmp-artifact-")
+	if err != nil {
+		return zero, "", fmt.Errorf("unable to create temporary artifact: %v", err)
+	}
+	defer tmpFile.Close()
+
+	tmpLocation := tmpFile.Name()
+
+	_, err = tmpFile.Write(newArtifactContent)
+	if err != nil {
+		return zero, tmpLocation, fmt.Errorf("unable to store temporary artifact: %v", err)
+	}
+
+	//Make sure that data has been written to disk
+	if err := tmpFile.Sync(); err != nil {
+		return zero, tmpLocation, fmt.Errorf("unable to sync file on disk: %v", err)
+	}
+
+	if err := perms.RestrictAccessToUser(tmpLocation); err != nil {
+		return zero, tmpLocation, fmt.Errorf("unable to set permission to temporary artifact: %v", err)
+	}
```
Here you are creating a file, writing a secret in it, and then setting permissions, so an attacker could read it in between... Can you restrict permissions before writing ?

misteriaud replied:
`os.CreateTemp` create the file with mode `0o600` which gives access only to the user that run the Agent.

misteriaud committed:
change the order of create (wait to get the lock to create the artifact) and increase the retryDelay time

misteriaud committed:
This commit improve the concurrent FS writter helper by:
- symplifying the ArtifactBuilder interface by merging two methods into one
- allowing multiple users to try locking the same lock file (in the case a root create the lock and dd-agent try to lock it afterward)
- implementing a system to cleaning up the lockfile when the artifact have been successfuly generated and stored

misteriaud committed:
Update auth_token and IPC_cert helper funcions to conform to new version of ArtiFactory interface
Remove the dcaAuthToken struct because double use with authToken struct

misteriaud committed:
Allow the user to set a custom timeout for Auth artifact LoadOrCreate initialization step

misteriaud committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into misteriaud/ASCII-2580-fix-race-condition-in-auth-token-creation

misteriaud committed:
update authToken ArtiFactory implementation and DCA AuthToken getter to match previous behavior

pgimalac commented on pkg/api/util/util.go:
```diff
@@ -104,14 +105,41 @@ func CreateAndSetAuthToken(config model.Reader) error {
+	errs := make(chan error, 2)
+	var ipccert, ipckey []byte
+
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		var err error
+		token, err = pkgtoken.FetchOrCreateAuthToken(ctx, config)
+		if err != nil {
+			err = fmt.Errorf("error while creating or fetching auth token: %v", err.Error())
+		}
+		errs <- err
+	}()
+
+	go func() {
+		defer wg.Done()
+		var err error
+		ipccert, ipckey, err = cert.FetchOrCreateIPCCert(ctx, config)
+		if err != nil {
+			err = fmt.Errorf("error while creating or fetching IPC cert: %v", err.Error())
+		}
+		errs <- err
+	}()
+
+	for i := 0; i < 2; i++ {
+		err := <-errs
+		if err != nil {
+			cancel()
+			wg.Wait()
+			return err
+		}
```
Is it really worth it having this kind of complexity to handle a case which should never happen (blocking 10s when trying to create a file) ?

misteriaud replied:
Indeed, I removed it

pgimalac commented on pkg/config/setup/config.go:
```diff
@@ -1099,6 +1099,8 @@ func agent(config pkgconfigmodel.Setup) {
 	// used to override the path where the IPC cert/key files are stored/retrieved
 	config.BindEnvAndSetDefault("ipc_cert_file_path", "")
+	// used to override the acceptable duration in seconds for the agent to load or create auth artifacts (auth_token and IPC cert/key files)
+	config.BindEnvAndSetDefault("auth_init_timeout", 10)
```
Be careful that here when you do `GetDuration("auth_init_timeout")` it will return 10ns, so you have to multiply with `time.Second` either when setting the default value or every time you get it, which is a footgun...
You could use `auth_init_timeout_second` to make it clear it's a number of seconds

misteriaud replied:
You're completely right, I opted to set the default to `10*time.Second` and not modifying the code reading the value. This will let more latitude to users.

pgimalac commented on pkg/api/util/util_test.go:
```diff
@@ -55,6 +61,162 @@ func TestIsIPv6(t *testing.T) {
+	err = CreateAndSetAuthToken(mockConfig)
+	duration := time.Since(start)
+	assert.Error(t, err)
+	assert.LessOrEqual(t, duration, mockConfig.GetDuration("auth_init_timeout")+time.Second)
```
I think ?
```suggestion
	assert.LessOrEqual(t, duration, mockConfig.GetDuration("auth_init_timeout")*time.Second)
```

misteriaud replied:
the `+` operator was on purpose, it's to compensate uncertainty on time measurment (e.g. the deadline is supposed to be set to 10s but in reality the time measured will be 10.00032s, so I prefered to put it to 11s).

pgimalac commented on pkg/api/util/util.go:
```diff
@@ -104,14 +105,41 @@ func CreateAndSetAuthToken(config model.Reader) error {
-	ipccert, ipckey, err := cert.CreateOrFetchAgentIPCCert(config)
-	if err != nil {
-		return err
+	ctx, cancel := context.WithTimeout(context.Background(), config.GetDuration("auth_init_timeout"))
```
You must multiply with `time.Second`

misteriaud replied:
I choosed to change the default value instead

GustavoCaso replied:
Should we move the creation of the timeout context closer to the function that actually use it? 

We create it at: util.CreateAndSetAuthToken -> security.FetchOrCreateAuthToken -> filesystem.FetchOrCreateArtifact -> tryLockContext

Could we create it on `filesystem.FetchOrCreateArtifact` that way it is easier to understand where the timeout might be coming from

misteriaud replied:
There is no blocking actions (such as FS read/write, syscalls) between the context creation and the beginning of the `filesystem.FetchOrCreateArtifact`, so it seems very unlikely that the this part of the execution have a non-negligeable impact on the overall remained time by the context for the artifact creation.
I was also wondering if we should create a context per artifact creation, but I think it's way more easier to set an overall timeout. 
Keeping in mind that the timeout default value is 10 seconds, if the auth_token FetchOrCreate call takes so much time that it doesn't let enough for the IPC_cert one would already be considered as problematic IMO.

pgimalac commented on pkg/api/util/util.go:
```diff
@@ -186,8 +214,11 @@ func InitDCAAuthToken(config model.Reader) error {
 		return nil
 	}
 
+	ctx, cancel := context.WithTimeout(context.Background(), config.GetDuration("auth_init_timeout"))
```
Same as above, multiply with `time.Second` (or change the default)

misteriaud replied:
Same as other comment, I decided to change the default value instead.

misteriaud committed:
improving error wrapping

misteriaud committed:
simplifying logic in auth_token and ipc_cert initialization

misteriaud committed:
fixing wrong auth_init_timeout config default value

misteriaud committed:
removing useless check in unit_test

clamoriniere reviewed (approved):
/lgtm for @DataDog/container-platform  files

misteriaud committed:
fix small issue on error return

pgimalac commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,210 @@
+// It copy the behavior of flock.TryLock() but retry if the lock have the wrong permissions.
+// This allow the case where a the lock file is created by a different user than the one trying to acquire the lock.
+// and let the opportunity to the other process to set the right permissions in the meantime.
+func TryLockContext(ctx context.Context, lockfile *flock.Flock, retryDelay time.Duration) (bool, error) {
```
That could probably be unexported

pgimalac reviewed (approved):
LGTM ! Would be nice to have more eyes on it though !

misteriaud committed:
minor error return update and set TryLockContext to private

misteriaud commented:
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

GustavoCaso commented on .gitignore:
```diff
@@ -64,6 +64,10 @@ datadog-agent*_amd64.deb
 # Ignore pem created during the tests
 *.pem
+
+# Ignore lock files created during tests
+*.lock
```
Should we clean the lock files instead?

misteriaud replied:
That's a good question ! I think since we are supposed to clean it at the end of the process we can remove it indeed.

GustavoCaso commented on pkg/api/security/cert/cert_generator.go:
```diff
@@ -47,29 +47,35 @@ func certTemplate() (*x509.Certificate, error) {
-		return nil, nil, err
+		return Certificate{}, err
 	}
 
 	rootKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
 	if err != nil {
-		return nil, nil, fmt.Errorf("Unable to generate IPC private key: %v", err)
+		return Certificate{}, fmt.Errorf("Unable to generate IPC private key: %v", err)
 	}
 
 	certDER, err := x509.CreateCertificate(rand.Reader, rootCertTmpl, rootCertTmpl, &rootKey.PublicKey, rootKey)
 	if err != nil {
-		return nil, nil, err
+		return Certificate{}, err
 	}
 
 	certPEM := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: certDER})
 	rawKey, err := x509.MarshalECPrivateKey(rootKey)
 	if err != nil {
-		return nil, nil, fmt.Errorf("Unable to marshall private key: %v", err)
+		return Certificate{}, fmt.Errorf("Unable to marshall private key: %v", err)
 	}
 
 	keyPEM := pem.EncodeToMemory(&pem.Block{Type: "EC PRIVATE KEY", Bytes: rawKey})
 
-	return certPEM, keyPEM, nil
+	return Certificate{certPEM, keyPEM}, nil
```
Absolutely nitpick: What about creating an instance of certificate at the beginning of the function `cert := Certificate{}`  and return that instead of `return Certifcate{}, ...`

The last line could be:

```go

cert.cert = certPEM
cert.key =  keyPEM
return cert, nil
```

misteriaud replied:
Both approachs seems equals to me  but I feel that it's a bit more explicit to return an default struct rather than a variable in this case. I don't have strong arguments though 😅

GustavoCaso commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,210 @@
+// and let the opportunity to the other process to set the right permissions in the meantime.
+func tryLockContext(ctx context.Context, lockfile *flock.Flock, retryDelay time.Duration) (bool, error) {
+	for {
+		if ok, err := lockfile.TryLock(); (ok || err != nil) && !errors.Is(err, fs.ErrPermission) {
```
Looking at the condition, I'm wondering if it should be like this instead:  `(ok || (err != nil && !errors.Is(err, fs.ErrPermission))` The way I read is, was the lock acquire or there was an error and the error wasn't because of wrong permissions, then return. 

I think the way is written now, works as well, but when is `ok` we still check if the error is a error permission, which might not be necessary

misteriaud replied:
I agree, your suggestion is also more readable ! Thanks

GustavoCaso commented on pkg/util/filesystem/concurrent_write_test.go:
```diff
@@ -0,0 +1,166 @@
+	assert.NoError(t, err)
+	assert.Equal(t, mockFactory.data, artifact)
+
+	// Verify the artifact file was created
```
I believe we are not removing the artifact for this test 😄

GustavoCaso commented on pkg/util/filesystem/concurrent_write_test.go:
```diff
@@ -0,0 +1,166 @@
+
+	results := make(chan string, numGoroutines)
+
+	// Start multiple goroutines to call LoadOrCreateArtifact in parallel
```
```suggestion
	// Start multiple goroutines to call FetchOrCreateArtifact in parallel
```

GustavoCaso commented on pkg/api/util/util_test.go:
```diff
@@ -55,6 +61,162 @@ func TestIsIPv6(t *testing.T) {
+	require.NoError(t, err)
+	defer lockFile.Unlock()
+	defer os.Remove(ipcCertFileLocation + ".lock")
+	mockConfig.SetWithoutSource("ipc_cert_file_path", ipcCertFileLocation)
```
We call this twice in the same test. Is that on purpose?

misteriaud replied:
Indeed, I remove the line, thanks

GustavoCaso commented on pkg/api/util/util_test.go:
```diff
@@ -55,6 +61,162 @@ func TestIsIPv6(t *testing.T) {
+	assert.EqualValues(t, createdServerTLSConfig.Certificates, GetTLSServerConfig().Certificates)
+}
+
+// This test check that if one of thje artifacts creation blocks, the function timeout
```
```suggestion
// This test check that if one of the artifacts creation blocks, the function timeout
```

GustavoCaso commented on pkg/api/util/util_test.go:
```diff
@@ -55,6 +61,162 @@ func TestIsIPv6(t *testing.T) {
+	// Reinitialize global vars to check with SetAuthToken
+	reinitGlobalVars()
+
+	// Check that CreateAndSetAuthToken returns no error
```
```suggestion
	// Check that SetAuthToken returns no error
```

GustavoCaso commented on pkg/api/util/util_test.go:
```diff
@@ -55,6 +61,162 @@ func TestIsIPv6(t *testing.T) {
+	require.NoError(t, err)
+	defer lockFile.Unlock()
+	defer os.Remove(ipcCertFileLocation + ".lock")
+	mockConfig.SetWithoutSource("ipc_cert_file_path", ipcCertFileLocation)
```
Called twice on this test 😄

GustavoCaso commented on pkg/api/util/util_test.go:
```diff
@@ -55,6 +61,162 @@ func TestIsIPv6(t *testing.T) {
+	defer os.Remove(ipcCertFileLocation + ".lock")
+	mockConfig.SetWithoutSource("ipc_cert_file_path", ipcCertFileLocation)
+
+	// Check that CreateAndSetAuthToken times out when the auth_token file is locked
```
```suggestion
	// Check that CreateAndSetAuthToken times out when the auth_token file is not writable
```

GustavoCaso commented on pkg/api/util/util_test.go:
```diff
@@ -55,6 +61,162 @@ func TestIsIPv6(t *testing.T) {
+	defer os.RemoveAll(authTokenDir)
+	mockConfig.SetWithoutSource("auth_token_file_path", path.Join(authTokenDir, "auth_token"))
+
+	// Create a lock file to simulate contention on ipc_cert_file
+	ipcCertFileLocation := path.Join(tmpDir, "ipc_cert_file")
+	mockConfig.SetWithoutSource("ipc_cert_file_path", ipcCertFileLocation)
+	lockFile := flock.New(ipcCertFileLocation + ".lock")
+	err = lockFile.Lock()
+	require.NoError(t, err)
+	defer lockFile.Unlock()
+	defer os.Remove(ipcCertFileLocation + ".lock")
+	mockConfig.SetWithoutSource("ipc_cert_file_path", ipcCertFileLocation)
```
Looking at the code, if the `FetchOrCreateAuthToken` function fails, we do not proceed to the creation of the IPC cert. Do we still need to create the IPC file lock?

misteriaud replied:
In fact it makes me realized that this test is no longer relevant, because artifact creation is sequencial.
For the context, I made a version where `CreateAndSetAuthToken` was trying to create `auth_token` and `ipc_cert` concurrently to minimize the time consumption.
But @pgimalac rightly suggested that the complexity that it introduce  is not worthing the gain in time in the happy path. So we get back to the sequencial version.

GustavoCaso reviewed (commented):
Amazing work @misteriaud. I left a few comments, looking forward to your answer

misteriaud committed:
Update tests:
- util_test.go: removing test which was relevant only when CreateAndSetAuthToken was call LoadOrCreate function concurrently for auth_token and ipc_cert
- concurrent_write_test.go: little on comment and useless assertion on existing tests, adding a test to assert that LoadOrCreate keep trying to lock the file even if it get permission denied

misteriaud committed:
Minor update:
 - removing `.lock` files from gitignore
 - improve tryLockContext stopping condition for better readability

misteriaud committed:
Adding debug logging for troubleshoot purpose:
- in LoadOrCreate function: logging whent the artifact creation succeed
- in pkg/api/util functions: logging a small format hash of the used auth stack (auth_token + ipc_cert)

dustmop reviewed (approved):
LGTM for agent-configuration. I have some small comments as well, but nothing major.

misteriaud commented on pkg/api/security/cert/cert_getter.go:
```diff
@@ -37,92 +35,44 @@ func GetCertFilepath(config configModel.Reader) string {
-	return fetchAgentIPCCert(config, true)
+func (certificateFactory) Generate() (Certificate, []byte, error) {
+	cert, err := generateCertKeyPair()
+	return cert, bytes.Join([][]byte{cert.cert, cert.key}, []byte{}), err
```
This change would modify the underlying value of `cert.cert` by making it contain both `cert.cert` and `cert.key`, isn't it ?
We don't want to modify `cert` struct.

misteriaud committed:
remove unused variable declaration

misteriaud committed:
modify behavior to try to read the artifact on every iteration and keep trying until defined timeout

misteriaud committed:
adding logging to notify that the auth stack loading started

misteriaud commented:
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

misteriaud committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into misteriaud/ASCII-2580-fix-race-condition-in-auth-token-creation

misteriaud commented:
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

misteriaud committed:
replace golang.org/x/exp/rand by crypto/rand in test

misteriaud commented:
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

misteriaud committed:
go mod tidy all

misteriaud commented:
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

pgimalac commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,207 @@
+	Deserialize([]byte) (T, error)
+}
+
+// FetchArtifact attempts to load an artifact using the provided factory.
+// If the artifact does not exist, it return an error.
+func FetchArtifact[T any](location string, factory ArtifactBuilder[T]) (T, error) {
+	var zero T
+
+	// Read the artifact
+	content, err := os.ReadFile(location)
+	if err != nil {
+		return zero, fmt.Errorf("unable to read artifact: %s", err)
+	}
+
+	// Try to load artifact
+	res, err := factory.Deserialize(content)
+	return res, err
+}
```
Should this function retry on failure, similarly to `FetchOrCreateArtifact` ?

misteriaud replied:
I renamed this function to `TryFetchArtifact` to make it clearer that it's non-blocking.
`CreateOrFetchArtifact` is expected to take some time because it is usually performed during the Agent's initialization step.
In contrast, `FetchAuthToken` is used in many places and, from my observations, it is not expected to be blocking until either the artifact has been retrieved or the context is done.
Ensuring that this change does not break anything will take more time, so it will be addressed in a separate PR.

pgimalac commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,207 @@
+	var zero T
+	var succeed bool
+
+	perms, err := NewPermission()
+	if err != nil {
+		return zero, log.Errorf("unable to init NewPermission: %v", err)
+	}
```
This could be moved below, in the happy path where the artifact already exists you don't need to do that

misteriaud replied:
Indeed !

pgimalac commented on pkg/util/filesystem/concurrent_write.go:
```diff
@@ -0,0 +1,207 @@
+
+	var lockErr error
+
+	// trying to read artifact or locking file
+	for {
+		// First check if another process were able to create and save artifact during wait
+		res, err := FetchArtifact(location, factory)
+		if err == nil {
+			succeed = true
+			return res, nil
+		}
+
+		// Trying to acquire lock
+		ok, err := fileLock.TryLock()
+		if err != nil {
+			lockErr = err
+			log.Debugf("unable to acquire lock: %v", err)
+		}
+		if ok {
+			break
+		}
+
+		select {
+		case <-ctx.Done():
+			return zero, errors.Join(fmt.Errorf("unable to read the artifact or acquire the lock in the given time"), lockErr)
+		case <-time.After(retryDelay):
+			// try again
+		}
+	}
```
We only retry locking / reading, but not writing, should we ?

misteriaud replied:
We should only retry elements that could fail due to actions from other Agent processes (e.g., failure to lock a file because another Agent has the lock, or failure to read the file because the Agent with the lock has not created it yet). The file creation step is not expected to fail because the Agent creating the artifact has obtained the lock, making it the only actor allowed to create it at that moment. If the file creation step fails, it is considered a fatal error and is not expected to be recoverable by external actions.

misteriaud committed:
reoarganize perm initialization in FetchOrCreate function and add function TryFetchArtifact to match better the retry behavior of FetchOrCreateArtifact function

pgimalac reviewed (approved):
LGTM

misteriaud committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into misteriaud/ASCII-2580-fix-race-condition-in-auth-token-creation

misteriaud commented:
/merge
